Core/include/Acts/EventData/VectorMultiTrajectory.hpp

```diff
@@ -0,0 +1,187 @@
+  }
+
+  ConstTrackStateProxy::Covariance covariance(IndexType parIdx) const override {
+    return ConstTrackStateProxy::Covariance{m_cov[parIdx].data()};
+  }
+
+  TrackStateProxy::Covariance jacobian(IndexType parIdx) override {
+    return TrackStateProxy::Covariance{m_jac[parIdx].data()};
+  }
+
+  ConstTrackStateProxy::Covariance jacobian(IndexType parIdx) const override {
+    return ConstTrackStateProxy::Covariance{m_jac[parIdx].data()};
+  }
+
+  TrackStateProxy::Measurement measurement(IndexType parIdx) override {
+    return TrackStateProxy::Measurement{m_meas[parIdx].data()};
+  }
+
+  ConstTrackStateProxy::Measurement measurement(
+      IndexType parIdx) const override {
+    return ConstTrackStateProxy::Measurement{m_meas[parIdx].data()};
+  }
+
+  TrackStateProxy::MeasurementCovariance measurementCovariance(
+      IndexType parIdx) override {
+    return TrackStateProxy::MeasurementCovariance{m_measCov[parIdx].data()};
+  }
+
+  ConstTrackStateProxy::MeasurementCovariance measurementCovariance(
+      IndexType parIdx) const override {
+    return ConstTrackStateProxy::MeasurementCovariance{
+        m_measCov[parIdx].data()};
+  }
+
+  std::size_t addTrackState(TrackStatePropMask mask = TrackStatePropMask::All,
+                            size_t iprevious = kNoPrevious) override;
+
+  void shareFrom(IndexType iself, IndexType iother,
+                 TrackStatePropMask shareSource,
+                 TrackStatePropMask shareTarget) override;
+
+  void unset(TrackStatePropMask target, IndexType istate) override;
+
+  bool has(HashedString key, IndexType istate) const override;
+
+  std::size_t size() const override { return m_index.size(); }
+
+  void clear() override;
+
+  struct DynamicColumnBase {
+    virtual ~DynamicColumnBase() = 0;
+
+    virtual void* get(size_t i) = 0;
+    virtual const void* get(size_t i) const = 0;
+
+    virtual void add() = 0;
+    virtual void clear() = 0;
+  };
+
+  template <typename T>
+  struct DynamicColumn : public VectorMultiTrajectory::DynamicColumnBase {
+    ~DynamicColumn() override = default;
+
+    void* get(size_t i) override {
+      assert(i < m_vector.size() && "DynamicColumn out of bounds");
+      return &m_vector[i];
+    }
+
+    const void* get(size_t i) const override {
+      assert(i < m_vector.size() && "DynamicColumn out of bounds");
+      return &m_vector[i];
+    }
+
+    void add() override { m_vector.emplace_back(); }
+    void clear() override { m_vector.clear(); }
+
+    std::vector<T> m_vector;
+  };
+
+ protected:
+  void* componentImpl(HashedString key, IndexType istate) override;
+
+  const void* componentImpl(HashedString key, IndexType istate) const override;
+
+  template <typename T>
+  constexpr void addColumnImpl(HashedString key) {
+    m_dynamic.insert({key, std::make_unique<DynamicColumn<T>>()});
+  }
+
+  constexpr bool hasColumnImpl(HashedString key) {
+    using namespace Acts::HashedStringLiteral;
+    switch (key) {
```

Comment:
That's a good point. I haven't benchmarked the dynamic columns bit at all yet (benches of the existing workflow/access pattern will be added to this PR shortly). 

As it stands, the number of dynamic columns is not known at compile time, since the class does not hold on to them, only in this factory method are they given as template params. We could think about adding such a mechanism though (i.e. having VectorMultiTrajectory or xAODMultiTrajectory accept dynamic columns as template parameters which it retains, then we can have a compile time check here). 

For right now, I think I would focus on the non-dynamic column performance, however.